README.md
# mat_v4
This library provides a C++ interface to write data in MATLAB mat v4 binary format to a file.
   * [License](LICENSE.md)
   * [Changelog](CHANGELOG.md)
   * [Contributing guide](CONTRIBUTING.md)

## Installation
CMake is used to build this library, which is exported as a library target called *mat_v4*. The header is added as:

```
#include "mat_v4/mat_v4.h"
```

The library can be also be compiled stand-alone using the CMake idiom of creating a *build* directory and then, from within that directory issuing:

```
cmake ..
make
```

This will build the library and an example executable called *mat_example*. The example executable source files are located at *examples/mat_example.cc*.

## Documentation
Additional documentation on the MATLAB mat file format is located in the *docs* folder.

## Namespace

This library is within the namespace *bfs*

## Functions

**void MatWrite(std::string name, const DATA &ref, FILE &ast;file)** This function writes data to the file given a MATLAB variable name, a reference to the data, and a pointer to the file. Data can be a matrix from the [Eigen library](http://eigen.tuxfamily.org), an array, or a vector. Static and dynamic matrices are supported and, in the case of a matrix, the matrix dimensions are used to determine the dimensions of the written data. Arrays and vectors are always written as row vectors. The data types supported are: double, float, int32_t, int16_t, uint16_t, and uint8_t.

```C++
  /* Output file */
  FILE *file = fopen("test.mat", "w");
  /* A dynamic matrix of doubles */
  Eigen::MatrixXd m(3, 3);
  m(0, 0) = 0.1;  m(0, 1) = 0.4;  m(0, 2) = 0.7;
  m(1, 0) = 0.2;  m(1, 1) = 0.5;  m(1, 2) = 0.8;
  m(2, 0) = 0.3;  m(2, 1) = 0.6;  m(2, 2) = 0.9;
  bfs::MatWrite("m", m, file);
  /* An array of uint8_t */
  std::array<uint8_t, 6> n;
  for (std::size_t i = 0; i < 6; i++) {
    n[i] = i;
  }
  bfs::MatWrite("n", n, file);
  /* A vector of floats */
  std::vector<float> o;
  for (std::size_t i = 0; i < 10; i++) {
    o.push_back(static_cast<float>(i) / 10);
  }
  bfs::MatWrite("o", o, file);
```


CHANGELOG.md
# Changelog

## v2.0.1
- Updated to eigen v2.0.0

## v2.0.0
- Updated namespace to *bfs*
- Renamed *Write* to *MatWrite*

## v1.0.3
- Updated CONTRIBUTING
- Switched to github as source for eigen

## v1.0.2
- Updated CONTRIBUTING
- Switched from ssh to https for *fetch_content*

## v1.0.1
- Updated license to MIT

## v1.0.0
- Initial baseline


include/mat_v4/mat_v4.h
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2021 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#ifndef INCLUDE_MAT_V4_MAT_V4_H_
#define INCLUDE_MAT_V4_MAT_V4_H_

#include <stdio.h>
#include <cstdint>
#include <string>
#include <array>
#include <vector>
#include "Eigen/Core"
#include "Eigen/Dense"

namespace bfs {

/* Write double matrices, arrays, and vectors */
template<int M, int N>
void MatWrite(std::string name, const Eigen::Matrix<double, M, N> &ref,
              FILE *file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 0,
    .mrows = static_cast<uint32_t>(ref.rows()),
    .ncols = static_cast<uint32_t>(ref.cols()),
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(double), header.mrows * header.ncols, file);
}
template<std::size_t M>
void MatWrite(std::string name, const std::array<double, M> &ref, FILE *file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 0,
    .mrows = static_cast<uint32_t>(M),
    .ncols = 1,
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(double), header.mrows * header.ncols, file);
}
void MatWrite(std::string name, const std::vector<double> &ref, FILE* file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 0,
    .mrows = static_cast<uint32_t>(ref.size()),
    .ncols = 1,
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(double), header.mrows * header.ncols, file);
}
/* Write float matrices, arrays, and vectors */
template<int M, int N>
void MatWrite(std::string name, const Eigen::Matrix<float, M, N> &ref,
              FILE *file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 10,
    .mrows = static_cast<uint32_t>(ref.rows()),
    .ncols = static_cast<uint32_t>(ref.cols()),
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(float), header.mrows * header.ncols, file);
}
template<std::size_t M>
void MatWrite(std::string name, const std::array<float, M> &ref, FILE *file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 10,
    .mrows = static_cast<uint32_t>(M),
    .ncols = 1,
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(float), header.mrows * header.ncols, file);
}
void MatWrite(std::string name, const std::vector<float> &ref, FILE* file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 10,
    .mrows = static_cast<uint32_t>(ref.size()),
    .ncols = 1,
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(float), header.mrows * header.ncols, file);
}
/* Write int32_t matrices, arrays, and vectors */
template<int M, int N>
void MatWrite(std::string name, const Eigen::Matrix<int32_t, M, N> &ref,
              FILE *file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 20,
    .mrows = static_cast<uint32_t>(ref.rows()),
    .ncols = static_cast<uint32_t>(ref.cols()),
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(int32_t), header.mrows * header.ncols, file);
}
template<std::size_t M>
void MatWrite(std::string name, const std::array<int32_t, M> &ref, FILE *file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 20,
    .mrows = static_cast<uint32_t>(M),
    .ncols = 1,
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(int32_t), header.mrows * header.ncols, file);
}
void MatWrite(std::string name, const std::vector<int32_t> &ref, FILE* file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 20,
    .mrows = static_cast<uint32_t>(ref.size()),
    .ncols = 1,
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(int32_t), header.mrows * header.ncols, file);
}
/* Write int16_t matrices, arrays, and vectors */
template<int M, int N>
void MatWrite(std::string name, const Eigen::Matrix<int16_t, M, N> &ref,
              FILE *file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 30,
    .mrows = static_cast<uint32_t>(ref.rows()),
    .ncols = static_cast<uint32_t>(ref.cols()),
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(int16_t), header.mrows * header.ncols, file);
}
template<std::size_t M>
void MatWrite(std::string name, const std::array<int16_t, M> &ref, FILE *file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 30,
    .mrows = static_cast<uint32_t>(M),
    .ncols = 1,
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(int16_t), header.mrows * header.ncols, file);
}
void MatWrite(std::string name, const std::vector<int16_t> &ref, FILE* file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 30,
    .mrows = static_cast<uint32_t>(ref.size()),
    .ncols = 1,
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(int16_t), header.mrows * header.ncols, file);
}
/* Write uint16_t matrices, arrays, and vectors */
template<int M, int N>
void MatWrite(std::string name, const Eigen::Matrix<uint16_t, M, N> &ref,
              FILE *file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 40,
    .mrows = static_cast<uint32_t>(ref.rows()),
    .ncols = static_cast<uint32_t>(ref.cols()),
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(uint16_t), header.mrows * header.ncols, file);
}
template<std::size_t M>
void MatWrite(std::string name, const std::array<uint16_t, M> &ref,
              FILE *file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 40,
    .mrows = static_cast<uint32_t>(M),
    .ncols = 1,
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(uint16_t), header.mrows * header.ncols, file);
}
void MatWrite(std::string name, const std::vector<uint16_t> &ref, FILE* file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 40,
    .mrows = static_cast<uint32_t>(ref.size()),
    .ncols = 1,
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(uint16_t), header.mrows * header.ncols, file);
}
/* Write uint8_t matrices, arrays, and vectors */
template<int M, int N>
void MatWrite(std::string name, const Eigen::Matrix<uint8_t, M, N> &ref,
              FILE *file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 50,
    .mrows = static_cast<uint32_t>(ref.rows()),
    .ncols = static_cast<uint32_t>(ref.cols()),
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(uint8_t), header.mrows * header.ncols, file);
}
template<std::size_t M>
void MatWrite(std::string name, const std::array<uint8_t, M> &ref, FILE *file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 50,
    .mrows = static_cast<uint32_t>(M),
    .ncols = 1,
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(uint8_t), header.mrows * header.ncols, file);
}
void MatWrite(std::string name, const std::vector<uint8_t> &ref, FILE* file) {
  /* Mat header */
  struct Header {
    uint32_t type;
    uint32_t mrows;
    uint32_t ncols;
    uint32_t imagf;
    uint32_t namelen;
  } header = {
    .type = 50,
    .mrows = static_cast<uint32_t>(ref.size()),
    .ncols = 1,
    .imagf = 0,
    .namelen = static_cast<uint32_t>(name.length() + 1)
  };
  /* Write the header */
  fwrite(&header, sizeof(header), 1, file);
  /* Write the name */
  fwrite(name.c_str(), sizeof(char), header.namelen, file);
  /* Write the data */
  fwrite(ref.data(), sizeof(uint8_t), header.mrows * header.ncols, file);
}

};  // namespace bfs

#endif  // INCLUDE_MAT_V4_MAT_V4_H_


examples/mat_example.cc
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2021 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#include <iostream>
#include <stdio.h>
#include "mat_v4/mat_v4.h"
#include "Eigen/Core"
#include "Eigen/Dense"

int main() {
  /* Output file */
  FILE *file = fopen("test.mat", "w");
  /* Matrix */
  Eigen::MatrixXd m(3, 3);
  m(0, 0) = 0.1;  m(0, 1) = 0.4;  m(0, 2) = 0.7;
  m(1, 0) = 0.2;  m(1, 1) = 0.5;  m(1, 2) = 0.8;
  m(2, 0) = 0.3;  m(2, 1) = 0.6;  m(2, 2) = 0.9;
  bfs::MatWrite("m", m, file);
  /* Array */
  std::array<uint8_t, 6> n;
  for (std::size_t i = 0; i < 6; i++) {
    n[i] = i;
  }
  bfs::MatWrite("n", n, file);
  /* Vector */
  std::vector<float> o;
  for (std::size_t i = 0; i < 10; i++) {
    o.push_back(static_cast<float>(i) / 10);
  }
  bfs::MatWrite("o", o, file);
}
